### DISCUSSION TITLE
Fabric and TurboModule support

### DISCUSSION THREAD

**Discussion by naftalibeder** [rnmapbox/maps#2106] [General]
This is a discussion on our Fabric/Turbo module migration.
See PR  #2819 for POC and concrete details.

mfazekas

----

This is an outline of what it would take to upgrade to Fabric and TurboModules (the new React Native renderer and native module interface).

I think it's too big of a job for just me, so I'd like to see if there's any interest in collaborating on it.

cc @mfazekas 

- Update dependencies
  - Update the `RNMapbox` library from 0.67.0 [(link)](https://react-native-community.github.io/upgrade-helper/?from=0.67.0&to=0.69.4)
  - Update the `Example` project from 0.68.1 [(link)](https://react-native-community.github.io/upgrade-helper/?from=0.68.1&to=0.69.4)
- Create spec to support code generation
  - TurboModule, e.g. `NativeRNMapbox.ts` [(link)](https://reactnative.dev/docs/new-architecture-library-intro#turbomodules)
  - Fabric components, e.g. `CameraNativeComponent.ts` [(link)](https://reactnative.dev/docs/new-architecture-library-intro#fabric-components)
- Migrate codebase to new API
  - Migrating findNodeHandle / getting a HostComponent [(link)](https://reactnative.dev/docs/new-architecture-library-intro#migrating-findnodehandle--getting-a-hostcomponent)
  - ~~Migrating .measure*()~~ Not used?
  - Migrating off setNativeProps [(link)](https://reactnative.dev/docs/new-architecture-library-intro#migrating-off-setnativeprops)
  - Move the call to `requireNativeComponent` to a separate file [(link)](https://reactnative.dev/docs/new-architecture-library-intro#move-the-call-to-requirenativecomponent-to-a-separate-file)
  - Migrating from `dispatchViewManagerCommand` to `codegenNativeCommands` [(link)](https://reactnative.dev/docs/new-architecture-library-intro#migrating-off-dispatchviewmanagercommand)
  - Update native method interfaces [(link)](https://reactnative.dev/docs/new-architecture-library-intro#updating-native-implementation)
- Enable native code generation
  - Android [(link)](https://reactnative.dev/docs/new-architecture-library-android)
  - iOS [(link)](https://reactnative.dev/docs/new-architecture-library-ios)

**Comment by systemlevel** [comment]
hi @naftalibeder, let's talk a bit more. Would you be interested in being sponsored to undertake this work? Would that free up your time to dedicate to this?

**Reply by naftalibeder** [reply]
Thanks for the offer! I would certainly prefer to just collaborate - do you have any availability at all? If not, I could probably pull off 4-5 hours a week, if you want to contact me at my email (in profile). Either way we should loop in @mfazekas so we don't do this in a way that contradicts his philosophy for the repo.

**Comment by mfazekas** [comment]
@naftalibeder I agree that we should try adopting TurboModules. 
- I'm ok if that is v10 only. (so we don't need to deal with MapLibre or old MapboxGL)
- Ideally we should start with a simple component, and continue based on that. 

Also there is https://docs.expo.dev/modules/module-api/ as I understood they also promise fabric/new architecture support but with much less ceremony (Swift + Kotlin DSL instead of C++ generated code) - https://www.youtube.com/watch?v=te7bwfNsRKg&t=4217s
But I'm not sure how mature, but sounds promising and much simpler than the TurboModule system

**Reply by naftalibeder** [reply]
That looks really cool - it does seem like maybe a lot to ask for people to install [another package](https://docs.expo.dev/bare/installing-expo-modules/) in order to install RNMapbox, but that's up to you. I think my inclination is to follow the current RN guidelines (Fabric/TurboModules).

But we should make a decision so we can get started - what are you feeling?

**Reply by mfazekas** [reply]
> That looks really cool - it does seem like maybe a lot to ask for people to install [another package](https://docs.expo.dev/bare/installing-expo-modules/) in order to install RNMapbox, but that's up to you. I think my inclination is to follow the current RN guidelines (Fabric/TurboModules).
> 
> But we should make a decision so we can get started - what are you feeling?

Yes you're right about requiring `expo-modules` might be problematic. Also I think we'll need to autogenerate some part of code from style spec, so we might be simpler just autogenerating the 'ts', instead of both kotlin and swift.

I'm fine with going with turbo modules.

**Reply by naftalibeder** [reply]
Ok great.

My guess is this will be a pretty big diff. Would you consider allowing this to be a major, partially breaking release - Mapbox v10 only (no Maplibre or pre-v10 Mapbox)? That would make it easier to support Fabric but keep it optional. I'm starting to worry about maintainability - there would be

- Mapbox or Maplibre
- If Mapbox, <v10 or v10
- If Mapbox v10, Paper or Fabric

Gets a little crazy.

**Reply by mfazekas** [reply]
@naftalibeder I plan to drop maplibre and mapboxgl as soon as we've released a stable v10. We're getting closer, but not there yet. Fabric should be optional for now for sure. Let's dicuss if with concrete case regarding fabric, where maintaining the old maplibre/mapboxgl is a pain.

**Reply by mfazekas** [reply]
FWIW, see this PR, to track POC for fabric: https://github.com/rnmapbox/maps/pull/2819

**Reply by alaindimabuyo** [reply]
commenting so I can be notified, for now I cant use the v10 of mapbox because of the fabric issue on android still using maplibre on android and mapbox v10 on ios so please dont drop maplibre if v10 is not as stable as it is on the new architecture thank you!

**Reply by janbaykara** [reply]
@mfazekas that PR is exciting! Curious what timelines look like on this and what could unblock progress, whether that be sponsorship or something else.

**Reply by mfazekas** [reply]
@janbaykara yes having my time spent on this sponsored, would be a great help, I cannot really justify spending my free time on this. Also last time I've checked the react-native codegen has various limitations we'd have to work around that.

**Reply by janbaykara** [reply]
OK good to know. It would be cool if people or orgs who are willing to sponsor this work would shout out in the comments and maybe raise some awareness that this is a fundable project. CC'ing @systemlevel who previously sounded like they might be willing to given [their last comment](https://github.com/rnmapbox/maps/discussions/2106#discussioncomment-3354098). I wonder if the Mapbox team would also be interested in contributing funds?

**Reply by mfazekas** [reply]
As an update #3022, was merged into main. `10.1*` releases should include initial fabric support. MapView should work but Sources/Layers and a lot of other stuff aren't going to work yet. Thanks much @j-piasecki
